src/core/utils/flash.cpp

```diff
@@ -29,54 +29,360 @@
 #include "flash.hpp"
 
 #include <stdio.h>
+#include <string.h>
 
 #include <openthread/platform/flash.h>
 
 #include "common/code_utils.hpp"
+#include "common/crc16.hpp"
 #include "common/instance.hpp"
 
 #if OPENTHREAD_CONFIG_PLATFORM_FLASH_API_ENABLE
 
 namespace ot {
 
-const uint32_t ot::Flash::sSwapActive;
-const uint32_t ot::Flash::sSwapInactive;
+#if OPENTHREAD_CONFIG_FLASH_LEGACY_COMPATIBILITY
+enum
+{
+    kFormatV1        = 0,
+    kFlashWordSizeV1 = 4, // in bytes
+};
+#endif
+
+enum
+{
+    kFormatV2        = 1,
+    kFlashWordSizeV2 = 8, // in bytes
+};
+
+OT_TOOL_PACKED_BEGIN
+class SwapHeader
+{
+public:
+    inline void Init(uint16_t aEraseCounter)
+    {
+        mMarker       = kMarkerInit;
+        mEraseCounter = aEraseCounter;
+        mReserved     = 0xffff;
+    }
+
+    inline void Load(otInstance *aInstance, uint8_t aSwapIndex)
+    {
+        otPlatFlashRead(aInstance, aSwapIndex, 0, this, sizeof(*this));
+    }
+
+    inline void Save(otInstance *aInstance, uint8_t aSwapIndex) const
+    {
+        otPlatFlashWrite(aInstance, aSwapIndex, 0, this, sizeof(*this));
+    }
+
+    bool IsActive(void) const { return (mMarker & kActiveMask) == kActive; }
+
+    void SetInactive(void)
+    {
+        mMarker       = ~mMarker;
+        mEraseCounter = 0xffff;
+        mReserved     = 0xffff;
+    }
+
+    uint8_t GetSize(void) const
+    {
+        return
+#if OPENTHREAD_CONFIG_FLASH_LEGACY_COMPATIBILITY
+            GetFormat() == kFormatV1 ? sizeof(mMarker) :
+#endif
+                                     sizeof(*this);
+    }
+
+    uint8_t GetFormat(void) const { return kFormatTop - (mMarker >> kFormatShift); }
+
+    uint16_t GetEraseCounter(void) const { return mEraseCounter; }
+
+private:
+    enum : uint32_t
+    {
+        kActive = 0x0e5cc5ee, // This matches the last 28 bits of the swap header format v1.
+
+        kActiveMask = 0x0fffffff,
+
+        kFormatTop   = 0xb, // This matches the first nibble of the swap header format v1.
+        kFormatShift = 28,
+
+        kMarkerInit = ((uint32_t)(kFormatTop - kFormatV2) << kFormatShift) | kActive,
+    };
+    static_assert(kActive >> kFormatShift == 0, "wrong kActive");
+    static_assert(kActiveMask == (uint32_t)((1 << kFormatShift) - 1), "wrong kActiveMask");
+
+    uint32_t mMarker;
+    uint16_t mEraseCounter;
```

Comment:
Are we only interested in the first 65k erase cycles? I think it is common for flash components to support 100k flash cycles, for example.